07_memory/usr_mem_test.c

```diff
@@ -0,0 +1,140 @@
+#include <stdio.h>
+#include <time.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <malloc.h>
+
+#define NSEC_IN_SEC	1000000000LL
+#define MAX_SIZE	64
+
+#define POW2(X)     (1ULL << (X))
+
+uint64_t get_diff(const struct timespec *t_start, const struct timespec *t_end)
+{
+	return  ((t_end->tv_sec*NSEC_IN_SEC) + t_end->tv_nsec) - 
+		((t_start->tv_sec*NSEC_IN_SEC) + t_start->tv_nsec);
+}
+
+char* mem_func_name[] = {"malloc", "calloc", "alloca"};
+
+struct live_mem_time {
+	uint64_t alloc_time;
+	uint64_t free_time;
+};
+
+enum alloc_func_name {
+	MALLOC = 0,
+	CALLOC,
+	ALLOCA,
+	NUM_FN
+};
+
+enum test_case {
+	TWO_POW_X = 0,
+	TWO_POW_X_PLUS_ONE,
+	NUM_TC
+};
+
+int get_time(uint64_t size, struct live_mem_time *t_live_mem, enum alloc_func_name func)
+{
+	struct timespec t_start, t_stop;
+	char *mem_ptr;
+
+	clock_gettime(CLOCK_MONOTONIC, &t_start);
+
+	switch(func) {
+		case MALLOC: 
+			mem_ptr = malloc(size * sizeof(*mem_ptr)); 
+			break;
+		case CALLOC:
+			mem_ptr = calloc(size, sizeof(*mem_ptr));
+			break;
+		case ALLOCA:
+			mem_ptr = alloca(size * sizeof(*mem_ptr));
+			break;
+		default: mem_ptr = NULL;
+		        break;
+	}
+
+	clock_gettime(CLOCK_MONOTONIC, &t_stop);
+
+	if (mem_ptr == NULL)
+		return 1;
+
+	t_live_mem->alloc_time = get_diff(&t_start, &t_stop);
+
+	if (func != ALLOCA) {
+		clock_gettime(CLOCK_MONOTONIC, &t_start);
+		free(mem_ptr);
+		clock_gettime(CLOCK_MONOTONIC, &t_stop);
+		t_live_mem->free_time = get_diff(&t_start, &t_stop);
+	}
+
+	return 0;
+}
+
+
+void run_test_mem(enum test_case test_num, enum alloc_func_name func)
+{
+	struct live_mem_time t_current = {0};
+	struct live_mem_time t_average = {0};
+	uint64_t size = 0;
+	uint64_t min = 0;
+	uint64_t max = 0;
+	uint64_t sum_alloc = 0;
+	uint64_t sum_free = 0;
+	int i = 0;
+
+	printf("====== Run memory test allocation! ============================\n");
+	printf("|\n");
+	printf("| function call: %s\n", mem_func_name[func]);
+	printf("| test case:     %s\n", (test_num)?"2^x + 1":"2^x");
+	printf("|\n");
+	printf("===============================================================\n");
+
+	for (i = 0; i < MAX_SIZE; i++) {
+
+		printf("| Id: %d\n", i);
+
+		size = POW2(i);
+
+		if(get_time(size, &t_current, func)) {
+			printf("| ### Memory cannot be allocated!\n");
+			break;
+		}
+
+		sum_alloc += t_current.alloc_time;
+		sum_free += t_current.free_time;
+
+		printf("| alloc time: %ld(ns)\t", t_current.alloc_time);
+		printf("free time: %ld(ns)\n", t_current.free_time);
+	}
+
+	t_average.alloc_time = sum_alloc/MAX_SIZE;
+	t_average.free_time = sum_free/MAX_SIZE;
+
+	printf("| Average alloc time: %ld(ns)\t", t_average.alloc_time);
+	printf("average free time: %ld(ns)\n", t_average.free_time);
```

Comment:
Please calculation average time for each size. For implementation repeat memory allocation few times for the same size.

Reply:
Done.